core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/async/ByteBuffersAsyncRequestBody.java

```diff
@@ -0,0 +1,136 @@
+ * using static methods on {@link AsyncRequestBody}
+ *
+ * @see AsyncRequestBody#fromBytes(byte[])
+ * @see AsyncRequestBody#fromByteBuffer(ByteBuffer)
+ * @see AsyncRequestBody#fromByteBuffers(ByteBuffer...)
+ * @see AsyncRequestBody#fromString(String)
+ */
+@SdkInternalApi
+public final class ByteBuffersAsyncRequestBody implements AsyncRequestBody {
+    private static final Logger log = Logger.loggerFor(ByteBuffersAsyncRequestBody.class);
+
+    private final String mimetype;
+    private final Long length;
+    private final ByteBuffer[] buffers;
+
+    private ByteBuffersAsyncRequestBody(String mimetype, Long length, ByteBuffer... buffers) {
+        this.mimetype = mimetype;
+        this.length = length;
+        this.buffers = buffers;
+    }
+
+    @Override
+    public Optional<Long> contentLength() {
+        return Optional.ofNullable(length);
+    }
+
+    @Override
+    public String contentType() {
+        return mimetype;
+    }
+
+    @Override
+    public void subscribe(Subscriber<? super ByteBuffer> s) {
+        // As per rule 1.9 we must throw NullPointerException if the subscriber parameter is null
+        if (s == null) {
+            throw new NullPointerException("Subscription MUST NOT be null.");
+        }
+
+        // As per 2.13, this method must return normally (i.e. not throw).
+        try {
+            s.onSubscribe(
+                new Subscription() {
+                    private final AtomicInteger index = new AtomicInteger(0);
+                    private final AtomicBoolean completed = new AtomicBoolean(false);
+
+                    @Override
+                    public void request(long n) {
+                        if (completed.get()) {
+                            return;
+                        }
+
+                        if (n > 0) {
+                            int i = index.getAndIncrement();
+
+                            if (i >= buffers.length) {
+                                return;
+                            }
+
+                            long remaining = n;
+
+                            do {
+                                ByteBuffer buffer = buffers[i];
+                                if (!buffer.hasArray()) {
+                                    buffer = ByteBuffer.wrap(BinaryUtils.copyBytesFrom(buffer));
+                                }
+                                s.onNext(buffer);
+                                remaining--;
+                            } while (remaining > 0 && (i = index.getAndIncrement()) < buffers.length);
+
+                            if (i >= buffers.length - 1 && completed.compareAndSet(false, true)) {
+                                s.onComplete();
+                            }
+                        } else {
+                            s.onError(new IllegalArgumentException("§3.9: non-positive requests are not allowed!"));
```

Comment:
`completed.set(true)` to drop future `request` calls?

Reply:
In hindsight I'm not sure if this would be correct, I lifted this logic from `ByteArrayAsyncRequestBody` so there's no logic change here, and I'm not sure if the this subscription would have to be marked as completed if it errors.
What do you think?

Reply:
I think it's correct, but this case probably never happens so I doubt we're testing it. If it's already like this, I'm not really opposed to leaving it like this. We need to move this to an easier-to-use publisher pattern anyway.